atlas-engineer/nyxt
# Nyxt URLs

This is a second shot at #1804. Less ambitious, more focused.

# What's New
- `nyxt:` URL scheme. 
  - Unencrypted.
  - Splits an evaluatable form into URL pieces e.g. `(list-buffers :cluster t)` turns into `nyxt:list-buffers?cluster=t&linear-view-p=t` for better parseability and nicer look in the address area.
    - The path (`list-buffers`) can only be one of registered command names (defined with `define-internal-page-command`).
    - The query parameters (`?cluster=t&linear-view-p=t`) can only be keyword of this command.
      - This keyword requirement severely restricts us in what args can we pass to the function, but ensures that argyments are easy to map to param names and helps in making `interactive` use easier.
      - Adding to that, param values can only be lisp constants. It's not really convenient, but it prevents something like `?cluster=(progn (hack-the-browser) t)&linear-view-p=t`.
- There's no more `with-current-html-buffer` -- it's replaced by `define-internal-page-command`.
- There are changes that hint at #1805 necessity: `style` of internal buffers is moved to `buffer` and `default-new-buffer-url` is moved to `browser`.
  - I'll duplicate that in #1805: it seems that `styleable-buffer` mix-in could come in handy.
  - Default new buffer URL semantically belongs to `browser`, because it should exist before any buffer can possibly exist.
- `lisp-url` is still there, merely as a deprecated placeholder to not cause compilation warnings. To be removed once #1762 is merged and JS->Lisp calls are supported.

# Things to Discuss
- Should we actually turn neat s-expressions into URLs? It feels like a proper way to analyse and transfer those, but there can be things I haven't considered in the transition. I need your input there :)

# How to Test
Compile Nyxt with this patch set and run it. Should work fine... except for side-effectful `lisp-url`s -- those are broken, and will stay this way until #1762 is merged with a better way of interactive Lisp evaluation.

What do you think of it?

Closes #1094
Closes #1728
Closes #1109
Closes #1091
